plugins/inputs/nsd/nsd.go

```diff
@@ -0,0 +1,200 @@
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/filter"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+type runner func(cmdName string, timeout internal.Duration, useSudo bool, server string) (*bytes.Buffer, error)
+
+type Nsd struct {
+	Binary  string
+	Timeout internal.Duration
+	UseSudo bool
+	Server  string
+
+	filter filter.Filter
+	run    runner
+}
+
+var defaultBinary = "/usr/sbin/nsd-control"
+var defaultTimeout = internal.Duration{Duration: time.Second}
+
+var sampleConfig = `
+  ## Address of server to connect to, read from nsd conf default, optionally ':port'
+  ## Will lookup IP if given a hostname
+  server = "127.0.0.1:8952"
+
+  ## If running as a restricted user you can prepend sudo for additional access:
+  # use_sudo = false
+
+  ## The default location of the nsd-control binary can be overridden with:
+  # binary = "/usr/sbin/nsd-control"
+
+  ## The default timeout of 1s can be overridden with:
+  # timeout = "1s"
+`
+
+func (s *Nsd) SampleConfig() string {
+	return sampleConfig
+}
+
+func (s *Nsd) Description() string {
+	return "A plugin to collect stats from the NSD DNS Server"
+}
+
+func nsdRunner(cmdName string, timeout internal.Duration, useSudo bool, server string) (*bytes.Buffer, error) {
+	cmdArgs := []string{"stats_noreset"}
+
+	if server != "" {
+		host, port, err := net.SplitHostPort(server)
+		if err != nil {
+			host = server
+			port = ""
+		}
+
+		resolver := net.Resolver{}
+		ctx, lookUpCancel := context.WithTimeout(context.Background(), timeout.Duration)
+		defer lookUpCancel()
+		serverIps, err := resolver.LookupIPAddr(ctx, host)
+		if err != nil {
+			return nil, fmt.Errorf("error looking up ip for server: %s: %s", server, err)
+		}
+		if len(serverIps) == 0 {
+			return nil, fmt.Errorf("error no ip for server: %s: %s", server, err)
+		}
+		server := serverIps[0].IP.String()
+		if port != "" {
+			server = server + "@" + port
+		}
+
+		cmdArgs = append([]string{"-s", server}, cmdArgs...)
+	}
+
+	cmd := exec.Command(cmdName, cmdArgs...)
+
+	if useSudo {
+		cmdArgs = append([]string{cmdName}, cmdArgs...)
+		cmd = exec.Command("sudo", cmdArgs...)
+	}
+
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	err := internal.RunTimeout(cmd, timeout.Duration)
+	if err != nil {
+		return &out, fmt.Errorf("error running nsd-control: %s (%s %v)", err, cmdName, cmdArgs)
+	}
+
+	return &out, nil
+}
+
+func (s *Nsd) Gather(acc telegraf.Accumulator) error {
+	out, err := s.run(s.Binary, s.Timeout, s.UseSudo, s.Server)
+	if err != nil {
+		return fmt.Errorf("error gathering metrics: %s", err)
+	}
+
+	fields := make(map[string]interface{})
+	fieldsServers := make(map[string]map[string]interface{})
+
+	host, port, err := net.SplitHostPort(s.Server)
+	// the port is nil since we already checked for other errors in the nsdRunner
+	if err != nil {
+		port = "8952"
+	}
+	_, localhostNet, _ := net.ParseCIDR("127.0.0.1/8")
```

Comment:
Make these package-wide variables and set them elsewhere (`init` perhaps) to avoid setting them every call to `Gather`